### ISSUE TITLE
Prediction on new data

### ISSUE THREAD

**Issue by phil0stine** [QingyongHu/RandLA-Net#56]
The paper mentions being able to directly process 3D point clouds without any pre-processing steps, but I am not sure how this would happen based on the implementation.

Let's say I have trained a model based on an existing dataset, but then want to run inference on a single brand new point cloud (not in the original datset) with no pre-processing, on a machine that does not have the original training/val data.

My questions are: 
- Can I run with a new point cloud of a slightly different dimension as the training set (I assume yes)
- Do I still need a KDTree for the input to run `get_batch_gen`?
- How would I setup the Network object to run inference directly on the cloud? Specifically, the `init_input_pipeline` function is required to setup the `flat_inputs` for the Network, but this function assumes that you can access the original training dataset. 

Thanks!

**Comment by QingyongHu** [comment]
Hi @phil0stine, thanks for your interest in our work!

First of all, the LFA module of our RandLA-Net is only operated in local neighboring sets, therefore, the learned weights of our RandLA-Net are agnostic to the number of points, i.e., the trained model is able to inference point clouds with different spatial size and number of points. In this case, KDTree is not required as you can directly feed all the points. 

But you are right, the preprocessing is indeed inevitable based on the current implementation, this is mainly because we use the tf.dataset to generate the data flow and use grid_subsampling to improve the performance.

If you want to take the trained model directly for inference, I would suggest that you still use the `feeddict ` mode in TensorFlow for training, and implement both random sampling and KNN search operations inside the graph, this would be a slightly slow, but would be flexible during inference. 

Hope this would be helpful, I will consider to re-implement the network in my free time.

**Comment by phil0stine** [comment]
Thanks for the help! I think I am close and am planning to use feed dict as
you suggest, but my problem is that I need to set up the inputs to the
graph in the Network constructor, but this depends on already having
processed data, and won't work with using a tf placeholder. If I can defer
setting the inputs until after the graph is set up and run then I think I
can do it

**Comment by phil0stine** [comment]
>  you still use the feeddict mode in TensorFlow for training, and implement both random sampling and KNN search operations inside the graph

Would the min working example be to keep CPU pre-processing, then change the Network constructor to not require known `xyz`, `features`, etc in the constructor? I originally thought I could use placeholders for  `xyz` and `features` (basically everything in `flat_inputs`), but things like `grid_subsampling` couldn't do the CPU ops on a `placeholder`. I guess this is why you suggest doing all that in the graph itself.

**Comment by QingyongHu** [comment]
Hi @phil0stine, sorry for the late response, as I was over-busy these days.
First, the function `grid_subsampling `is not necessary for our RandLA-Net, you can also directly operate on the raw point clouds, instead of the subsampled point clouds. In addition, `py_func `
is also a flexible way to use other functions. As for the neighbour index, pool index, upsample index, all of these can be calculated inside the graph.

**Comment by phil0stine** [comment]
Thanks, @QingyongHu , I appreciate it and plan to continue asking design questions for my understanding.  For example, in the implementation of the `building_block`, in the first pass of local spatial encoding you first get a relative position encoding, then use an MLP to convert the encoding to the desired shape for concatenation. In the second pass, you use another MLP but instead of using the original relative position encoding as input, you simply use the output of the first MLP. Do you think it would make a difference if you were to instead use the same input as the first pass but just convert it to a different shape?

**Comment by QingyongHu** [comment]
Hi @phil0stine, I have tried both, but the current one would be slightly better (~1%). This is not an important problem, you can use both implementations.

**Comment by phil0stine** [comment]
Excellent thanks! It does make for a slightly cleaner implementation

**Comment by phil0stine** [comment]
Hi @QingyongHu ; I want to make sure I am reading this correctly. Two questions in particular. In `tf_map`, the input `batch_pc_idx` values come from the initial random sampling of points in a given batch, `spatially_regular_gen`. Then, the layers are iterated and KNN / subsampling takes place.

First question is, you do not randomly reshuffle `x` for each layer, correct? I had assumed from the paper that you did, but I don't see how that happens here.

Secondly, am I correct to read that for each `pick_idx` during `spatially_regular_gen`, the output of `crop_pc` is a point cloud with ~50k (for KITTI) randomly shuffled nearest neighbors to the pick point?

**Comment by QingyongHu** [comment]
Hi @phil0stine , sorry for my late response.
1. Yes, we only shuffle once and get the top 1/4 index, we think this is also equivalent to random sampling in principle
2. Yes, we pick the same number of points from the point cloud for better parallelization. But in fact, you can change the number of points during inference if your point cloud is large than this

**Comment by hongocvuong1998** [comment]
So, Can I predict a list point for segmentation. I will input the model (1,46000,3) (batch_size, numper_point, value) and output(1,46000,4) (I assume I have 4 class)

**Comment by wakasu** [comment]
@phil0stine 
Hey there,
I was wondering how you implemented the prediction pipeline on unlabeled dataset.
I have trained the model and am able to run validation using the datasets provided but not able wrap my head around how can i run prediction on an unlabeled dataset.

Any help will be appreciated.

**Comment by phil0stine** [comment]
@wakasu 
Hi, I ended up taking a lot of the early advice, such as `the neighbour index, pool index, upsample index, all of these can be calculated inside the graph.`. Also, I didn't go with TF1 anymore just due to it being deprecated, so I would suggest either TF2 or PyTorch. There's a PyTorch [implementation](https://github.com/aRI0U/RandLA-Net-pytorch) that, while it is probably not exactly what you want, could get you started.

**Comment by wakasu** [comment]
@phil0stine 
Thanks for the pointers.
We want to see whether this model is feasible for real-time segmentation or not, that's why just want to do some quick inference.
If something comes up might trouble you again. lol

**Comment by escanillans** [comment]
@phil0stine 
Hi, I was wondering if you could give more guidance as to how you calculated the neighbour index, etc. inside the graph? For example, does that mean you placed those calculations within _RandLANet.py_?

**Comment by phil0stine** [comment]
Hi @escanillans for nearest neighbors I followed the example [PyTorch](https://github.com/aRI0U/RandLA-Net-pytorch/blob/master/model.py) approach, which does the KNN operation on the cpu. Hope it helps

**Comment by momei123** [comment]
Hello, is this problem solved? If you need to leave the contact information (I add you), I have done two new user data, and the results are perfectly presented.

**Comment by phil0stine** [comment]
@momei123 Sure I will close this, although I think some were hoping that this capability would be formally added as part of the repository. Instead, it seems like most of us had to write custom implementations.
